design-proposals/klip-26-java-client-interfaces.md

```diff
@@ -0,0 +1,715 @@
+   */
+  CompletableFuture<DdlResponse> executeDdl(String sql);
+
+  /**
+   * Execute DML statement: `CREATE STREAM ... AS SELECT`, `CREATE TABLE ... AS SELECT`, `INSERT INTO`
+   */
+  CompletableFuture<DmlResponse> executeDml(String sql);
+
+  /**
+   * Execute DML statement: `CREATE STREAM ... AS SELECT`, `CREATE TABLE ... AS SELECT`, `INSERT INTO`
+   */
+  CompletableFuture<DmlResponse> executeDml(String sql, Map<String, Object> properties);
+
+```
+with
+```
+public interface DdlResponse {
+
+  enum Status { QUEUED, PARSING, EXECUTING, RUNNING, TERMINATED, SUCCESS, ERROR }
+  
+  Status getStatus();
+  
+  String getStatusMessage();
+  
+  long getCommandSequenceNumber();
+
+  String getCommandId();
+}
+```
+and `DmlResponse` is identical.
+
+I don't love this proposal since `DDL` and `DML` aren't used throughout our docs but I'm not a huge fan of the alternatives I considered either, such as
+```
+  CompletableFuture<CreateSourceResponse> createStream(String sql);
+
+  CompletableFuture<CreateSourceResponse> createStream(String sql, Map<String, Object> properties);
+
+  CompletableFuture<CreateSourceResponse> createTable(String sql);
+
+  CompletableFuture<CreateSourceResponse> createTable(String sql, Map<String, Object> properties);
+
+  CompletableFuture<InsertIntoResponse> insertIntoSource(String sql);
+
+  CompletableFuture<InsertIntoResponse> insertIntoSource(String sql, Map<String, Object> properties);
+```
+
+In this version, the implementations of `createStream(...)` and `createTable(...)` would be identical and could be replaced with a single `createSource(...)`
+but the name of this method feels confusing. As with the previous proposal, `CreateSourceResponse` and `InsertIntoResponse` are identical.
+
+Other alternatives include separating `CREATE <STREAM/TABLE>` and `CREATE <STREAM/TABLE> ... AS SELECT` into separate methods, but that feels unnecessarily complex.
+
+#### `DROP <STREAM/TABLE>`
+
+`Client` methods:
+```
+  /**
+   * Drop stream. The underlying Kafka topic will not be deleted.
+   */
+  CompletableFuture<DropSourceResponse> dropStream(String streamName);
+
+  /**
+   * Drop stream. The underlying Kafka topic may optionally be deleted.
+   */
+  CompletableFuture<DropSourceResponse> dropStream(String streamName, boolean deleteTopic);
+
+  /**
+   * Drop table. The underlying Kafka topic will not be deleted.
+   */
+  CompletableFuture<DropSourceResponse> dropTable(String tableName);
+
+  /**
+   * Drop table. The underlying Kafka topic may optionally be deleted.
+   */
+  CompletableFuture<DropSourceResponse> dropTable(String tableName, boolean deleteTopic);
+```
+where `DropSourceResponse` is actually the same as `DdlResponse`/`DmlResponse`/`CreateSourceResponse`/`InsertIntoResponse` above.
+
+Again, the implementations of `dropStream(...)` and `dropTable(...)` would be the same so we could instead have a single `dropSource(...)`, but the naming might be confusing.
```

Comment:
again, not sure why we can't just group this into `executeStatement` instead of separating all these methods - seems straightforward to me and even more user friendly if I'm programatically constructing statements

Reply:
As above.